hack/lib/common.sh

```diff
@@ -37,3 +40,92 @@ function is_installed {
 function install_service_monitor_crd {
   kubectl apply -f https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.35/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml
 }
+
+# Skip fetching and untaring the tools by setting the SKIP_FETCH_TOOLS variable
+# in your environment to any value:
+#
+# $ SKIP_FETCH_TOOLS=1 ./test.sh
+#
+# If you skip fetching tools, this script will use the tools already on your
+# machine, but rebuild the operator-sdk binary.
+SKIP_FETCH_TOOLS=${SKIP_FETCH_TOOLS:-""}
+
+# prepare the e2e test staging dir, containing test tools (SKIP_FETCH_TOOLS aware).
+function prepare_staging_dir {
+
+  header_text "preparing staging dir $1"
+
+  if [[ -z "$SKIP_FETCH_TOOLS" ]]; then
+    rm -rf "$1"
+  else
+    rm -f "$1/bin/operator-sdk"
+  fi
+
+  mkdir -p "$1"
+}
+
+# Fetch k8s API gen tools and make it available under $1/bin.
+function fetch_tools {
+  if [[ -z "$SKIP_FETCH_TOOLS" ]]; then
+    fetch_envtest_tools $@
+    install_kind $@
+  fi
+}
+
+# Fetch tools required for envtest.
+function fetch_envtest_tools {
+
+  # TODO: make our own tarball containing envtest binaries: etcd, kubectl, kube-apiserver
+  #
+  # To get k8s server binaries:
+  # server_tar="kubernetes-server-$(go env GOOS)-$(go env GOARCH).tar.gz"
+  # url=https://dl.k8s.io/$K8S_VERSION/$server_tar
+  # curl -fL --retry 3 --keepalive-time 2 "${url}" -o "${tmp_sdk_root}/${server_tar}"
+  # tar -zxvf "${tmp_sdk_root}/${server_tar}"
+
+  local tools_archive_name="kubebuilder-tools-1.16.4-$(go env GOOS)-$(go env GOARCH).tar.gz"
```

Comment:
I guess I'm a little thrown off by the discrepancy between the k8s tools(etcd, kubectl, api-server) version vs the k8s version used by Kind.
But can we parametrize the version here to make it configurable, or highlight it a bit better.
```suggestion
  local tools_archive_name="kubebuilder-tools-$k8s_tools_version-$(go env GOOS)-$(go env GOARCH).tar.gz"
```

Reply:
If/whenever we do switch to using a local control plane(etcd, kube-apiserver) we'll want to source this version the same way we do the Kind cluster version i.e [K8S_VERSION](https://github.com/operator-framework/operator-sdk/blob/master/Makefile#L14).

Otherwise we'll end up testing against a `1.16.4` control-plane vs a `1.18.2`.

Reply:
I'll parameterize the tools version. We need to publish our own tools tarball fairly soon.

Reply:
Good catcher.

Reply:
Good catcher. But it has not 1.18.2. See: https://storage.googleapis.com/kubebuilder-tools/
kb ship etcd and etc. I think that it is related to it.